### PR TITLE
Fix: url.params fatal error: concurrent map read and map write

### PR DESCRIPTION
**Pull request by jack15083**
**What this PR does**:
fix url.params fatal error: concurrent map read and map write
**Which issue(s) this PR fixes**:
https://github.com/apache/dubbo-go/issues/772

### CODE REVIEW

File: common/url.go [hunk 1]
```diff
@@ -393,34 +396,33 @@ func (c URL) Service() string {
...
-	c.params.Add(key, value)
+	c.AddParam(key, value)
 }
 
 // SetParam will put the key-value pair into url
-// it's not thread safe.
-// think twice before you want to use this method
 // usually it should only be invoked when you want to initialized an url
 func (c *URL) SetParam(key string, value string) {
+	c.paramsLock.RLock()
```
Comment (AlexStocks) [inline common/url.go:417]: Hey, guy, pls using write lock instead.

File: common/url.go [hunk 2]
```diff
@@ -393,34 +396,33 @@ func (c URL) Service() string {
...
+	defer c.paramsLock.RUnlock()
 	c.params.Set(key, value)
 }
 
 // RangeParams will iterate the params
-// it's not thread-safe
 func (c *URL) RangeParams(f func(key, value string) bool) {
```
Comment (AlexStocks) [inline common/url.go:423]: u should exec the @f outside the lock to limit the lock scope.

```Go
        var (
            flag bool
            key string
            value string
        )
        func() {
        c.paramsLock.RLock()
	defer c.paramsLock.RUnlock()
	   for k, v := range c.params {
		if !f(k, v[0]) {
                         key = k
                         value = v[0]
                         flag = true
			break
		}
	  }
        }()
        if flag {
          f(key, value) 
        }
```

**Review by AlexStocks — APPROVED** [review]

File: common/url.go (hunk 1 above)
Comment (zouyx) [reply common/url.go:417]: Agree

File: common/url.go (hunk 2 above)
Comment (jack15083) [reply common/url.go:423]: test this code, u should remove if flag {f(key, value)}

File: common/url.go [hunk 3]
```diff
@@ -393,43 +396,59 @@ func (c URL) Service() string {
...
 // AddParamAvoidNil will add key-value pair
-// Not thread-safe
-// think twice before using it.
 func (c *URL) AddParamAvoidNil(key string, value string) {
 	if c.params == nil {
 		c.params = url.Values{}
 	}
 
-	c.params.Add(key, value)
+	c.AddParam(key, value)
```
Comment (zouyx) [inline common/url.go:411]: I think the write lock should cover all line of this method

**Comment by georgehao** [comment]
pls see #511 , because many url param of function is copy, just add rwmutex may appear warnings.

**Review by watermelo — APPROVED** [review]

**Review by fangyincheng — APPROVED** [review]

**Pull request merged**